RecoParticleFlow/PFClusterProducerAlpaka/plugins/alpaka/PFClusterProducerKernel.dev.cc

```diff
@@ -0,0 +1,2224 @@
+       this software without specific prior written permission.
+
+  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+  ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+  WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+  DISCLAIMED. IN NO EVENT SHALL TEXAS STATE UNIVERSITY BE LIABLE FOR ANY
+  DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+  Authors: Jayadharini Jaiganesh and Martin Burtscher
+
+  URL: The latest version of this code is available at
+  https://userweb.cs.txstate.edu/~burtscher/research/ECL-CC/.
+
+  Publication: This work is described in detail in the following paper.
+  Jayadharini Jaiganesh and Martin Burtscher. A High-Performance Connected
+  Components Implementation for GPUs. Proceedings of the 2018 ACM International
+  Symposium on High-Performance Parallel and Distributed Computing, pp. 92-104.
+  June 2018.
+*/
+
+namespace ALPAKA_ACCELERATOR_NAMESPACE {
+
+  using namespace cms::alpakatools;
+
+  using PFClustering::common::PFLayer;
+
+  constexpr const float PI_F = 3.141592654f;
+  static const int threadsPerBlockForClustering = std::is_same_v<Device, alpaka::DevCpu> ? 32 : 512;
+  static const int warpsize = 32;
+
+  // Preparation of topo inputs. Initializing topoId, egdeIdx, nEdges, edgeList
+  template <typename TAcc, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
+  ALPAKA_FN_ACC void prepareTopoInputs(const TAcc& acc,
+                                       const unsigned int nRH,
+                                       reco::PFRecHitHostCollection::ConstView pfRecHits,
+                                       reco::ClusteringVarsDeviceCollection::View clusteringVars,
+                                       reco::ClusteringEdgeVarsDeviceCollection::View clusteringEdgeVars,
+                                       uint32_t* __restrict__ nSeeds) {
+    if (alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[0u] == 0 &&
+        alpaka::getIdx<alpaka::Grid, alpaka::Blocks>(acc)[0u] == 0) {
+      clusteringVars.nEdges() = nRH * 8;
+      clusteringEdgeVars[nRH].pfrh_edgeIdx() = nRH * 8;
+    }
+
+    for (uint32_t i = alpaka::getIdx<alpaka::Grid, alpaka::Blocks>(acc)[0u] *
+                          alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc)[0u] +
+                      alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[0u];
+         i < nRH;
+         i += alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc)[0u] *
+              alpaka::getWorkDiv<alpaka::Grid, alpaka::Blocks>(acc)[0u]) {
```

Comment:
General question for all device-side loops in this file. They explicitly loop over the thread indices of the block. I would _guess_ that not having the element-level loop would prevent autovectorization on the CPU serial backend (if the compiler would otherwise be able to vectorize). Would any of the looping utilities in `HeterogeneousCore/AlpakaInterface` be useful here?